### DISCUSSION TITLE
Simultaneous buttons events

### DISCUSSION THREAD

**Discussion by jsleroy** [bxparks/AceButton#83] [General]
Is it possible to detect simultaneous events? For example, assuming there is 3 buttons (A, B and C), is that possible or how would you detect that A and B are pressed together ?

**Comment by bxparks** [comment]
AceButton itself does not support simultaneous button events. Keeping each button independent of each other allows the code to be more understandable and maintainable. However, I think you can build a custom subclass of `IEventHandler` that may be sufficient in some cases.

The `IEventHandler` object can intercept the `kEventPressed` and `kEventReleased` from buttons A and B, and has internal variables to keep track of the states of both buttons. When it detects that both buttons are pressed, then you can trigger the desired action. Now, I can think of 2 interpretations of "simultaneous press":

1) You want to detect if both buttons are pressed, even if they are pressed many seconds apart. Example, you press and hold button A, then many seconds later, you press button B, then release button B, then press button B again. Do you want to detect 2 simultaneous button press events?
2) You want to detect a simultaneous press event *only* if both buttons are pressed within a short time interval of each other (e.g. 100 milliseconds). If they are pressed many seconds apart, nothing will happen.

The custom `IEventHandler` can handle both cases. I created a sample program called [examples/SimultaneousButtons](https://github.com/bxparks/AceButton/tree/develop/examples/SimultaneousButtons) that you can grab from the `develop` branch. The `ButtonHandler` class looks like this:

```C++
class ButtonHandler: public IEventHandler {
...
  public:
    explicit ButtonHandler(bool requireConcurrentTrigger = false)
        : mRequireConcurrentTrigger(requireConcurrentTrigger)
    {}

    void handleEvent(AceButton* button, uint8_t eventType,
        uint8_t /*buttonState*/) override {
      ...
    }

    void handleB1Pressed() {
      Serial.println("B1 pressed");
    }

    void handleB2Pressed() {
      Serial.println("B2 pressed");
    }

    void handleBothPressed() {
      Serial.println("B1 and B2 BOTH pressed");
    }
    ...
};
```

If you pass `requireConcurrentTrigger = false`, this corresponds to case (1) that I describe above where any simultaneous press is detected. If you pass `requireConcurrentTrigger = true`, this corresponds to case (2) above, where both buttons need to be pressed within 100 milliseconds of each other.

Finally, I want to draw attention to the edge case that I commented in the docstring for the `areBothPressed()` function:

```C++
    /**
     * Determine if both buttons are pressed at the same time.
     * 
     * If mRequireConcurrentTrigger is true, then also check if the Pressed
     * event of one button occurred within kConcurrentTriggerMillis of the other
     * button. The mB{x}PressedMillis is a 16-bit integer which can rollover
     * after 65536 milliseconds. So if the first button is held down for more
     * than (65536 - 100 = 65435) milliseconds, it is possible for the 2nd
     * button to trigger a concurrent press when pressed 65 seconds after the
     * first. I can think of 2 possible fixes:
     *
     *  1) Enable kEventRepeatPress, and use one of those events to reset
     *     the mB{x}PressedMillis of the other button appropriately.
     *  2) Use an external periodic task to reset the mB{x}PressedMillis
     *     every few seconds.
     */
    bool areBothPressed() {
      ...
    }
```

The internal time stamps are `uint16_t`, which rolls over after 65536 milliseconds. If you hold down the first button for about 65 seconds, then it is possible for the 2nd button to trigger an invalid simultaneous button event due to the rollover. The comment contains 2 possible solutions to this problem.

Hope this helps...
